Scripts/BuildDatabank/NMRPCA_timerelax.py

```diff
@@ -1,466 +1,673 @@
-    tail2_atoms = []
-
-    for key in atoms_dict:
-        atom = atoms_dict[key]['ATOMNAME']
-        if atom[0] == "H":
-            continue
-        if atoms_dict[key]['FRAGMENT'] == 'headgroup' or atoms_dict[key]['FRAGMENT'] == 'glycerol backbone':
-            head_atoms.append(atom)
-        elif atoms_dict[key]['FRAGMENT'] == 'sn-1':
-            tail1_atoms.append(atom)
-        elif atoms_dict[key]['FRAGMENT'] == 'sn-2':
-            tail2_atoms.append(atom)
-        heavy_atoms.append(atom)
-
-    heavy_top = u.select_atoms(f'resname {lipid_resname} and not name H*')
-    head_top = u.select_atoms(f'(name {" ".join(head_atoms)}) and resname {lipid_resname}')
-    tail1_top = u.select_atoms(f'(name {" ".join(tail1_atoms)}) and resname {lipid_resname}')
-    tail2_top = u.select_atoms(f'(name {" ".join(tail2_atoms)}) and resname {lipid_resname}')
-
-    n_lipid = head_top.n_residues          # number of lipids
-    
-    n_atoms_lipid = len(heavy_atoms)
-    n_atoms_head = len(head_atoms)
-    n_atoms_tail1 = len(tail1_atoms)
-    n_atoms_tail2 = len(tail2_atoms)
-
-
-    coords = np.empty((n_frames*n_lipid*n_atoms_lipid, 3))
-    for i, ts in enumerate(traj):
-        #coords[i*n_atoms_lipid*n_lipid:(i+1)*n_atoms_lipid*n_lipid, :] = framesLipCoord(n_lipid, n_atoms_head, head_top, n_atoms_tail1, tail1_top, n_atoms_tail2, tail2_top)
-        coords[i*n_atoms_lipid*n_lipid:(i+1)*n_atoms_lipid*n_lipid, :] = heavy_top.positions #heavy_trj.positions
-
-    coords = coords.reshape((n_frames,n_lipid,n_atoms_lipid,3))
-    coords = np.swapaxes(coords,0,1)
-    coords = coords.reshape((n_frames*n_lipid,n_atoms_lipid,3))
-
-    # new Universe (trajn) based on heavy atom positions
-    atom_resindex = np.repeat(0, n_atoms_lipid)
-    # [] are neaded
-    res_name = [lipid_resname]
-    trajn = mda.Universe.empty(n_atoms_lipid, 1, atom_resindex=atom_resindex, trajectory=True)
-    trajn.add_TopologyAttr('name', heavy_atoms)
-    trajn.add_TopologyAttr('resname', res_name)
-    trajn.load_new(coords)
-    
-    # average structure
-    av = align.AverageStructure(trajn, ref_frame=0).run()
-    avg_str = av.results.universe 
-    align.AlignTraj(trajn, avg_str).run() #aligning of the new trajectory along the average structure
-    
-    # positions from the average structure
-    avg_pos = avg_str.select_atoms('all').positions.astype(np.float64).\
-    reshape(1, avg_str.select_atoms('all').n_atoms * 3)
-    
-    # positions from trajn   
-    n_framesn = len(trajn.trajectory)
-    n_atoms_lipid = trajn.select_atoms('all').n_atoms
-    coords = np.empty((n_framesn*n_atoms_lipid, 3))
-    print(n_framesn, n_atoms_lipid)
-    for i, ts in enumerate(trajn.trajectory):
-        coords[i*n_atoms_lipid:(i+1)*n_atoms_lipid, :] = trajn.select_atoms('all').positions
-    return coords, avg_pos, n_lipid, n_frames
-
-###################################### AMBER ################################################################################################  
```

Comment:
def get_headtails_lipid(self, n_atoms_head, headlist, n_atoms_tail, taillist, ind):
        
        # Get coordinates of lipid head
        headcoords=np.empty((n_atoms_head, 3))
        headcoords[0:n_atoms_head, :]=headlist.positions[ind*n_atoms_head:(ind+1)*n_atoms_head]
        # Get coordinates of lipid tails
        tailcoords=[]
        for i in range(len(n_atoms_tail)):
            tailcoords.append([])
            tailcoords[i]=np.empty((n_atoms_tail[i], 3))
            tailcoords[i][0:n_atoms_tail[i], :]=taillist[i].positions[ind*n_atoms_tail[i]:(ind+1)*n_atoms_tail[i]]
        
        return headcoords, tailcoords
        
    def merger(self, n_lipids, n_atoms_head, headlist, n_atoms_tail, taillist):
        
        # merge coordinates of a lipid and get lipid coordinates within the frame
        mergecoords=np.empty((0,3))
        for i in range(n_lipids):
            headcoords,tailcoords=self.get_headtails_lipid(n_atoms_head, headlist, n_atoms_tail, taillist, i)
            mergecoords_lipid=headcoords
            for j in range(len(tailcoords)):
                mergecoords_lipid=np.concatenate((mergecoords_lipid, tailcoords[j]))
            
            mergecoords=np.concatenate((mergecoords, mergecoords_lipid))
        
        return mergecoords

    def concatenateTrajAmber(self):
        
        if self.lipid_resname == 'POPC':
            head, tails = 'PC', ['PA','OL']
            
        elif self.lipid_resname == 'POPG':
            head, tails = 'PGR', ['PA','OL']
            
        elif self.lipid_resname == 'POPS':
            head, tails = 'PS', ['PA','OL']
            
        elif self.lipid_resname == 'POPE':
            head, tails = 'PE', ['PA','OL']
        
        traj = self.traj.trajectory
        n_frames = len(traj)
        
        headlist = self.traj.select_atoms(f"not name H* and resname {self.head}")
        n_heads = headlist.n_residues
        n_atoms_head = headlist.n_atoms // n_heads
        
        taillist = []
        selres_tail = []
        n_tails = [] 
        n_atoms_tail = []
        
        n_atoms_lipid = n_atoms_head
        
        for i, j in enumerate(tails):
            
            taillist.append([])
            selres_tail.append([])
            n_tails.append([])
            n_atoms_tail.append([])
            
            # Find lipid tails that correspond to a particular head (based on the structure).
            taillist[i] = self.traj.select_atoms(f"(not name H* and resname {j}) and around 2.0 (resname {head} and not name H*)")
            selres_tail[i] = "(resid " + " or resid ".join(str(taillist[i].resids[:])[1:-1].split()) + ") and not name H*"
            taillist[i] = self.traj.select_atoms(selres_tail[i]) 
            
            n_tails[i] = taillist[i].n_residues
            n_atoms_tail[i] = taillist[i].n_atoms // n_heads
            n_atoms_lipid += n_atoms_tail[i]
        
        # Get all coordinates
        n_lipids = n_heads
        coords = np.empty((n_frames*n_lipids*n_atoms_lipid, 3))
        for i, ts in enumerate(traj):
            coords[i*n_atoms_lipid*n_lipids:(i+1)*n_atoms_lipid*n_lipids, :] = self.merger(n_lipids, n_atoms_head, headlist, n_atoms_tail, taillist)
            
        # Adding axis to separate individual lipid trajectories
        coords = coords.reshape((n_frames, n_lipids, n_atoms_lipid, 3))
        # Swapping time frame with lipid axis
        # Now we have continuous lipid trajectory
        coords = np.swapaxes(coords, 0, 1)
        # Reshaping to desired shape
        coords = coords.reshape((n_frames * n_lipids, n_atoms_lipid, 3))

        # Creating new Universe from the concatenated coordinates
        atom_resindex = np.repeat(0, n_atoms_lipid)
        concatenated_traj = mda.Universe.empty(
            n_atoms_lipid, 1, atom_resindex=atom_resindex, trajectory=True
        )
        concatenated_traj.add_TopologyAttr("resname", [self.lipid_resname])
        concatenated_traj.load_new(coords)

        return concatenated_traj, n_lipids, n_frames * n_lipids